### PR TITLE
Add list-metadata to list files in a directory with all their metadata

### PR DESCRIPTION
**Pull request by tao-xinxiu**
Add another component for the list request of dataspace to retrieve the full list of dirs and files in a directory with all their metadata.

### CODE REVIEW

File: rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/dto/ListFileMetadata.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    public void addPermissions(String filename, String permission) {
+        permissions.put(filename, permission);
+    }
+
+    public void addLastModifiedDates(String filename, Date date) {
+        lastModifiedDates.put(filename, date);
+    }
+
+    public void addSizes(String filename, long size) {
```
Comment (fviale) [inline rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/dto/ListFileMetadata.java:74]: addSize

File: rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/dto/ListFileMetadata.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    public void addType(String filename, String type) {
+        types.put(filename, type);
+    }
+
+    public void addPermissions(String filename, String permission) {
+        permissions.put(filename, permission);
+    }
+
+    public void addLastModifiedDates(String filename, Date date) {
```
Comment (fviale) [inline rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/dto/ListFileMetadata.java:70]: addLastModifiedDate

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/FileSystem.java [hunk 3]
```diff
@@ -149,6 +141,27 @@ public static ListFile list(FileObject fo, List<String> includes, List<String> e
         return answer;
     }
 
+    private static Map<String, FileObject> listFileObjects(FileObject fo, List<String> includes, List<String> excludes)
+            throws FileSystemException {
+        fo.refresh();
+        List<FileObject> fileObjects = Lists.newArrayList();
+        Map<String, FileObject> result = new HashMap<>();
```
Comment (fviale) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/dataspace/FileSystem.java:148]: new LinkedHashMap()

File: rest/rest-client/src/test/java/functionaltests/DataTransferTest.java [hunk 4]
```diff
@@ -494,6 +497,135 @@ public void testListFilesRecursiveWithPattern() throws Exception {
...
+        System.out.println("Metadata: " + listFile);
+        assertEquals("DIRECTORY", listFile.getTypes().get(TEMP_DIR_NAME));
+        assertNotNull(listFile.getLastModifiedDates().get(TEMP_DIR_NAME));
+        assertEquals("rwx", listFile.getPermissions().get(TEMP_DIR_NAME));
+        assertEquals("text/plain", listFile.getTypes().get(TEMP_FILE_TXT_NAME));
+        assertNotNull(listFile.getLastModifiedDates().get(TEMP_FILE_TXT_NAME));
+        if (checkExecutable) {
+            assertEquals("rw-", listFile.getPermissions().get(TEMP_FILE_TXT_NAME));
+        } else {
+            // on some system (e.g., windows), File.setExecutable may be failed, in this case, we don't check its exec permission.
```
Comment (fviale) [inline rest/rest-client/src/test/java/functionaltests/DataTransferTest.java:584]: on windows, all files have executable rights

**Review by aminelouati — APPROVED** [review]

**Pull request merged**